### ISSUE TITLE
R&S Editor UX allows multiple Beneficiaries per article

### ISSUE THREAD

**Issue by wesrowe** [department-of-veterans-affairs/va.gov-cms#13188]
### User story
**AS AN** Editor of Resources & Support node
**I WANT** to be able to add multiple Beneficiary types to an article
**SO THAT** the R&S page more accurately represents reality.

### Background
* Beneficiaries is a Taxonomy: https://prod.cms.va.gov/admin/structure/taxonomy/manage/audience_beneficiaries/overview
  * Also used on Campaign Landing Pages (Why This Matters > Audiences > Beneficiaries), but as an entity reference.  
* On Resources & Support, currently, it's used within the **"Audience and topics" paragraph type**, Example node: https://prod.cms.va.gov/node/53778/edit
* "Audience and topics" paragraph type
  * We can control the cardinality within a Paragraph, so anywhere the Paragraph used will be affected. 
  *  is used in 6 content types. As of 10/31/23:
      * [checklist](https://prod.cms.va.gov/admin/content?title=&type=checklist&moderation_state=All&owner=All) = 2 nodes, draft
      - [media_list_images](https://prod.cms.va.gov/admin/content?title=&type=media_list_images&moderation_state=All&owner=All) = 2 nodes, both draft
      - [media_list_videos](https://prod.cms.va.gov/admin/content?title=&type=media_list_videos&moderation_state=All&owner=All) = 2 nodes, both draft
      - [step_by_step](https://prod.cms.va.gov/admin/content?title=&type=step_by_step&moderation_state=All&owner=All) = 14 nodes, some published -- CAIA are editors. 
      - [faq_multiple_q_a](https://prod.cms.va.gov/admin/content?title=&type=faq_multiple_q_a&moderation_state=All&owner=All) = CAIA approved making multiple choice
      - [q_a](https://prod.cms.va.gov/admin/content?title=&type=q_a&moderation_state=All&owner=All) = CAIA approved making multiple choice
      - [support_resources_detail_page](https://prod.cms.va.gov/admin/content?title=&type=support_resources_detail_page&moderation_state=All&owner=All) = CAIA approved making multiple choice

Audience & Topics paragraph usage: 
<img width="329" alt="image" src="https://github.com/department-of-veterans-affairs/va.gov-cms/assets/4054752/d90fc2e6-c30a-4454-a5dc-e82f0cf7a1be">

## Engineering notes

The FE is capable of handling 1 or many tags for R&S pages at this time without modifications. The `support_resources_detail_page.drupal.liquid` template, which renders the related node types, leverages a common `tags.drupal.liquid` template for rendering tags, which itself leverages the Liquid filter `getTagsList`, which handles multiple instances of an audience tag:

<details>

<summary>
getTagsList code
</summary>

```
  liquid.filters.getTagsList = fieldTags => {
    const {
      entity: {
        fieldTopics = [],
        fieldAudienceBeneficiares,
        fieldNonBeneficiares,
      },
    } = fieldTags;

    const topics = fieldTopics.map(topic => ({
      ...topic.entity,
      categoryLabel: 'Topics',
    }));

    const audiences = [
      fieldAudienceBeneficiares?.entity,
      fieldNonBeneficiares?.entity,
    ]
      .filter(tag => !!tag)
      .map(audience => ({
        ...audience,
        categoryLabel: 'Audience',
      }));

    const tagList = [...topics, ...audiences];

    return _.sortBy(tagList, 'name');
  };
```

</details>

**TBD: find uses of the underlying field (fieldAudienceBeneficiares:field_audience_beneficiares) to ensure there are no other uses of this field outside of the above template.**

### Stakeholders / Decisions
**CAIA** has approved allowing multiple selections for CAIA content types.
**Facilities** We need to review and understand any impact to them with this change. 

### Next Steps
- [x] @jilladams @FranECross: Check with Facilities team
- [x] FE team: assess impact in FE templates, including CLP and 6 other types listed above

## Prefinement tasks
- [X] Audit CMS code for uses of this field
- [x] Audit FE code for uses of this field
  - [x] e.g. also what will happen on the FE?
- [x] Review findings with PO/PM to confirm we should make this change
- [X] Verify plan with CAIA
- [x] Verify plan with Facilities team
  - [x] Facilities team response: No evidence of the Audience & Topics paragraph type being referenced by any Facility content types (see screenshot in comment below https://github.com/department-of-veterans-affairs/va.gov-cms/issues/13188#issuecomment-1791511732

## Acceptance criteria
- [ ] For all 6 affected content types: Content editors see Beneficiary choices as checkboxes
  - [ ] Editors are able to choose 1 or more checkboxes as the audience target
  - [ ] Previously selected radio button will have its checkbox auto-checked
- [ ] Ensure that this field has adequate automated tests & existing unit tests are updated
- [ ] PR should be marked DO NOT MERGE and ticket moved to Complete Pending Integration until FE work is complete
- [ ] ~~KB article is updated~~ not needed, [comment](https://github.com/department-of-veterans-affairs/va.gov-cms/issues/13188#issuecomment-1828527487)

**Comment by wesrowe** [comment]
Needed to refine:
- What changes are required on FE? Does template use the Beneficiaries field?
- CMS design: are we switching to checkboxes (from radio buttons?)

**Comment by wesrowe** [comment]
Wes [slacked](https://dsva.slack.com/archives/C52CL1PKQ/p1684272000058349) Danielle. 

Daniel provided other compnents that would be impacted:
![image](https://github.com/department-of-veterans-affairs/va.gov-cms/assets/4054752/b2d5910d-9cc7-4aed-9f2d-280cd3cb5f44)

Danielle is okay with enabling multiple beneficiaries on these types, with the caveat that relevant help text could be nice.

> I think this is okay, but would love 
[@Mikki Northuis](https://dsva.slack.com/team/U1W46RJKA)
's thoughts as well. I do wonder if we could add some helper text to help guide folks a bit so they don't always choose a bunch of beneficiaries by default.

(Waiting for Mikki's response.)

**Comment by wesrowe** [comment]
Mikki replied: She's okay with the change. She speculated these types are all R&S, and cautioned that she might be concerned if any are used in Facilities products. PW to-do: can we audit or otherwise rule out that Facilities editors see any of these content types?

**Comment by jilladams** [comment]
Related weird old ticket to consider re: Facilities / how other administrations may think about this: #5732

**Comment by jilladams** [comment]
We know CLPs include this list, as Why This Matters > Audiences > Beneficiaries. https://prod.cms.va.gov/node/add/campaign_landing_page

**Comment by jilladams** [comment]
## Content types that would be impacted:
* checklist = 2 nodes, draft
*  media_list_images = 2 nodes, both draft
* media_list_videos = 2 nodes, both draft
* step_by_step = 2 nodes, both draft
* faq_multiple_q_a = CAIA approved making multiple choice
* q_a = CAIA approved making multiple choice
* support_resources_detail_page =  CAIA approved making multiple choice

## Next steps
* Decide if we want to isolate this to R&S only or go ahead modifying for all these content types. In order to decide: 
  * Need to understand FE impact. 
    * Do all of these have FE templates? (video list, image list, etc)
    * What will happen if the FE templates receive multiple Beneficiaries?
* Need to clarify with Facilities what Editor impact, and change management required.

**Comment by jilladams** [comment]
Reached out to Facilities team: https://dsva.slack.com/archives/C0FQSS30V/p1698880051537049

**Comment by FranECross** [comment]
Response from Facilities: No evidence of the Audience & Topics paragraph type being referenced by any Facility content types
![screencapture-staging-cms-va-gov-admin-reports-content-model-entity-diagram-paragraph-audience-topics-2023-11-01-18_44_15](https://github.com/department-of-veterans-affairs/va.gov-cms/assets/147188767/5505f50b-8c5d-4ab7-ac29-d07dac09d98d)

**Comment by FranECross** [comment]
@FranECross @jilladams  Need to find the KB(s) that need to be updated.

**Comment by FranECross** [comment]
@jilladams Thanks so much for this amazing detail. I'll create a ticket for updating the videos and will link to the epic.

**Comment by dsasser** [comment]
@FranECross @jilladams 

Some questions:

1.  Does this ticket pertain only to the 'Beneficiaries'? Or should the 'Non-Beneficiaries' also become checkboxes?
![Screenshot 2023-11-28 at 3 13 43 PM](https://github.com/department-of-veterans-affairs/va.gov-cms/assets/221539/c5e5dd54-c3b7-44eb-abee-edfc9ee0d745)


2. There are no existing tests at all for the Resources and Support content type. Adding them doesn't affect the effort, since we pointed at a 5 to account for unknown testing lift. However...


3. There is a mechanism controlling the total amount of "Tags" (the sum of all topics and beneficiaries/non-beneficiaries), which limits the total amount to 4. However, it either isn't behaving correctly, or needs modified help text. Modifying the help text might be possible in this ticket, providing all necessary stakeholders can weigh in within the sprint boundary, but  perhaps we should carve out a separate ticket to unpack the expected behavior if we think it might need to be updated.

![Screenshot 2023-11-28 at 3 25 55 PM](https://github.com/department-of-veterans-affairs/va.gov-cms/assets/221539/17e42316-3527-40b5-abd5-3aa4f3b9506a)

**Comment by FranECross** [comment]
@dsasser Thanks for all this info! I'll carve off one or two more tickets as you suggest so that the originally asked for change is within the sprint boundary, and that'll give me time to check on what to do with the Topics and Non-beneficiaries. Can this be in it's own branch so that we can hold back on releasing it until I can unpack the other requirements, etc.? Thanks again for all your work on this! cc @jilladams

**Comment by FranECross** [comment]
@dsasser Regarding updated the videos, here is the new ticket created: #16239   cc @jilladams

**Comment by jilladams** [comment]
CAIA are the stakeholders here so a question in the #sitewide-content-accessibility-ia channel ( i think I got the order right) is a good starting place. You can tag Danielle and Randi Hecht, who have I think been involved here.

**Comment by dsasser** [comment]
## Status Update 12/1/23

Work is done and approved, but we cannot merge until the [FE work](https://github.com/department-of-veterans-affairs/va.gov-cms/issues/16099) is ready.

**Comment by jilladams** [comment]
@dsasser clarifying: do CAIA's answers re: Topics / Audience changes block this? Or we could ship with what we've done and fast follow with any additional changes needed for that?

**Comment by dsasser** [comment]
> @dsasser clarifying: do CAIA's answers re: Topics / Audience changes block this? Or we could ship with what we've done and fast follow with any additional changes needed for that?

Good question, I neglected to indicate that prior. The questions are non-blocking for this issue.

**Comment by FranECross** [comment]
Thanks, @jilladams  and @dsasser . I'll follow up again on the questions and we can ship in the meantime. 🎉

**Comment by FranECross** [comment]
Followed up on original Slack message today: https://dsva.slack.com/archives/C01K37HRUAH/p1701705550448669?thread_ts=1701292198.711309&cid=C01K37HRUAH

**Comment by FranECross** [comment]
@dsasser We've heard back from Mikki/Danielle... 

1. Non-beneficiaries: leaving as radio button selection is fine (from Mikki: these are not really utilized in the experience at this time and not intended to be used in the near future.  No work on this is necessary.  In addition, those groups are more mutually exclusive, and I would say that the radio button is more accurate for those options.)
2. For Topics, do you @dsasser have a feel for how much work it would take to allow the user to select 1-4 topics, regardless of the number of Beneficiaries chosen?  Mikki's comment is below:

- Topics - the options available are far from complete and are not helpful at this time.  In order for these to be helpful 1) you should be able to select more than 1 and 2) we need a much more comprehensive list and likely a hierarchy of topics. If the work to allow selecting more than 1 is extensive, then this could be postponed for now.  cc @jilladams

**Comment by dsasser** [comment]
@FranECross 

I'm not sure where the idea that only one Topic can be selected arose, but it is possible to select 4 Topics at this time. Since there are only 4 of them, I'm not sure we have any work to do on that front until more possible Topics are added (if ever they are).

**Comment by FranECross** [comment]
It's in reference to your note and screenshots way above (pasted below) indicating that it doesn't seem to be working correctly....

There is a mechanism controlling the total amount of "Tags" (the sum of all topics and beneficiaries/non-beneficiaries), which limits the total amount to 4. However, it either isn't behaving correctly, or needs modified help text. Modifying the help text might be possible in this ticket, providing all necessary stakeholders can weigh in within the sprint boundary, but perhaps we should carve out a separate ticket to unpack the expected behavior if we think it might need to be updated.

**Comment by dsasser** [comment]
Apologies for the confusion @FranECross. The current behavior works by counting the total number of Tags, including Topics and Beneficiaries, with the intent to cap the total to 4. Once the editor makes a total of 4 selections, the Tags field is locked for further editing (I had forgotten this was the case, hence my confusion). In other words, it doesn't matter if Topics is 1 or more, but once a total of 4 is reached, Topics is locked.

If we don't care to lock the user to 4 selections any longer, I propose we:

1. Change the help text, removing the maximum of 4 statement.
2. Updating the code to not restrict/lock out the Topics field at all.

This is a small lift, probably a 2.

I hope that helps, and again, apologies that didn't recall that Topics were being restricted to 1 in the screenshot above leading to the understanding that a user cannot select more than that.

**Comment by FranECross** [comment]
No worries at all, @dsasser ! This is great info, and I'll create a secondary ticket to do as you suggest in 1 & 2 above. Thanks again!

**Comment by jilladams** [comment]
Noting: FE change shipped that will handle either shape of data (single, or unlimited cardinality). CMS PR merged, will deploy tomorrow. We'll likely need CAIA's help for a prod verification, POC tbd.

**Comment by FranECross** [comment]
@jilladams thanks for the follow up on this with Dave and here! Here's a slack thread where Mikki, Danielle and R Hecht offered feedback on the multiple beneficiaries. I'll be creating follow up tickets (thought I had already) to completely remove the restriction on how many tags/beneficiaries can be added (the code has them currently lumped together in counting). https://dsva.slack.com/archives/C01K37HRUAH/p1701292198711309

**Comment by jilladams** [comment]
This change is live in prod. We can edit prod nodes without CAIA to confirm, so closing, and we can address any add'l follow up via the next steps ticket Fran is going to file.

**Comment by FranECross** [comment]
Ticket created https://github.com/department-of-veterans-affairs/va.gov-cms/issues/16574
